### DISCUSSION TITLE
Creating backups of LLDAP installation

### DISCUSSION THREAD

**Discussion by pfaelzerchen** [lldap/lldap#794] [Q&A] [answered]
I've deployed LLDAP on Kubernetes, following the examples from https://github.com/Evantage-WS/lldap-kubernetes. Instead of sqlite I deployed a PostgreSQL database to store the data. So far, it works fine.

As backup, I dumped the postgresql database with

```bash
kubectl run db-postgresql-client --rm --tty -i --restart='Never' --namespace lldap \
  --image docker.io/bitnami/postgresql:16.1.0-debian-11-r16 --env="PGPASSWORD=$POSTGRES_PASSWORD" \
  --command -- pg_dump --host postgres-postgresql.lldap.svc.cluster.local -U postgres -d lldap > "$BACKUP_FILE"
```

I tested the recovery by (clean state, no old database/lldap pvcs or anything like that):

- Deploying the PostgreSQL database
- Restoring the dumped database file
- Deploying LLDAP

When I want to log in, I'll get wrong username or password for all accounts.

What did I forget?

**Comment by nitnelave** [comment, marked as answer]
Hmm, my first thought is that the private key used to encrypt the passwords was not backed up, and was regenerated (invalidating all the passwords).
I see that the Kubernetes guide sets up a persistent folder, there should be a "server_key" in there that you want to keep.

If you update your LLDAP to the latest version, it should tell you exactly what's going on if it's a private key problem. That would help make sure that it's the issue.

Usually for K8s, or even just in general, I recommend using a key seed rather than a key file. That way you can even get rid of the persistent directory (if you move all your config to environment variables).

**Comment by pfaelzerchen** [comment]
Got it. It is on the PV. And thanks for the hint with the env variables. Getting rid of useless volumes is always a good idea.

Just a question. The key_file seems binary. Any chance I can extract it to the key seed or do I have to start from scratch?

**Reply by nitnelave** [reply]
The opposite is possible: extracting a key file from a key seed. But the seed just initializes a RNG, so you can't reverse that.

**Reply by nitnelave** [reply]
And note that "start from scratch" is just for the passwords: the rest of the data is still there.

If you have trouble because you lost the admin password, recent LLDAP versions have an option to reset it on the command line

**Reply by pfaelzerchen** [reply]
Sorry, can you give me one last hint, what the command would be? I found `lldap_set_password`. But this one need the admin-password to set a password. I tried `--token` instead, but

```bash
/app # ./lldap_set_password -b http://localhost:17170/ --username admin --password password-123 --token "$LLDAP_JWT_SECRET"
Error: HTTP status client error (401 Unauthorized) for url (http://localhost:17170/auth/opaque/register/start)
```

**Reply by pfaelzerchen** [reply]
Sorry again, the obvious way was to send me password reset links.

**Comment by pfaelzerchen** [comment]
OK, one last thing: After I switched to a key seed, I could recover the admin password and log back in. When I tried to change the passwords of my bind users, I got an error message: Could not initiate password change

I had to recover them with password reset links, too. Do you have any idea, or is this probably a bug?

**Reply by nitnelave** [reply]
How did you try to set the passwords? In the web UI, logged in as admin? Or with lldap_set_password?

Either way, it should be possible, that sounds like a bug.

**Comment by pfaelzerchen** [comment]
I tried it through the web frontend. But I recognized that my firefox is not able to send password reset links, konqueror does. Probably some side effect of privacy addons. Could have prevented this, too. I'll try to investigate on it and open a bug report, if necessary.

Thanks for your help so far. I really like LLDAP. Together with authelia it's a really good solution for my homelab.
